src/app/clusters/actions-server/actions-server.cpp

```diff
@@ -0,0 +1,355 @@
+{
+    ChipLogProgress(Zcl, "ActionsServer: OnActionFailed");
+
+    // Record ActionFailed event
+    EventNumber eventNumber;
+    Events::ActionFailed::Type event{ actionId, invokeId, actionState, actionError };
+
+    if (CHIP_NO_ERROR != LogEvent(event, endpoint, eventNumber))
+    {
+        ChipLogError(Zcl, "ActionsServer: Failed to record OnActionFailed event");
+    }
+}
+
+CHIP_ERROR Instance::ReadActionListAttribute(const AttributeValueEncoder::ListEncodeHelper & encoder)
+{
+    if (GetInstance()->mDelegate == nullptr)
+    {
+        ChipLogError(Zcl, "Actions delegate is null!!!");
+        return CHIP_ERROR_INCORRECT_STATE;
+    }
+    for (uint16_t i = 0; i <= kMaxActionList; i++)
+    {
+        ActionStructStorage action;
+
+        CHIP_ERROR err = GetInstance()->mDelegate->ReadActionAtIndex(i, action);
+        if (err == CHIP_ERROR_PROVIDER_LIST_EXHAUSTED)
+        {
+            return CHIP_NO_ERROR;
+        }
+
+        ReturnErrorOnFailure(encoder.Encode(action));
+    }
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR Instance::ReadEndpointListAttribute(const AttributeValueEncoder::ListEncodeHelper & encoder)
+{
+    if (GetInstance()->mDelegate == nullptr)
+    {
+        ChipLogError(Zcl, "Actions delegate is null!!!");
+        return CHIP_ERROR_INCORRECT_STATE;
+    }
+    for (uint16_t i = 0; i <= kMaxEndpointList; i++)
+    {
+        EndpointListStorage epList;
+
+        CHIP_ERROR err = GetInstance()->mDelegate->ReadEndpointListAtIndex(i, epList);
+        if (err == CHIP_ERROR_PROVIDER_LIST_EXHAUSTED)
+        {
+            return CHIP_NO_ERROR;
+        }
+
+        ReturnErrorOnFailure(encoder.Encode(epList));
+    }
+    return CHIP_NO_ERROR;
+}
+
+void Instance::SetDefaultDelegate(Delegate * aDelegate)
+{
+    if (aDelegate == nullptr)
+    {
+        ChipLogError(Zcl, "Cannot set empty delegate!!!");
+        return;
+    }
+    mDelegate = aDelegate;
+}
+
+CHIP_ERROR Instance::Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
+{
+    VerifyOrDie(aPath.mClusterId == Actions::Id);
+
+    switch (aPath.mAttributeId)
+    {
+    case ActionList::Id: {
+        Instance * d = this;
+        return aEncoder.EncodeList([d](const auto & encoder) -> CHIP_ERROR { return d->ReadActionListAttribute(encoder); });
+    }
+    case EndpointLists::Id: {
+        Instance * d = this;
+        return aEncoder.EncodeList([d](const auto & encoder) -> CHIP_ERROR { return d->ReadEndpointListAttribute(encoder); });
+    }
+    default:
+        break;
+    }
+    return CHIP_NO_ERROR;
+}
+
+bool Instance::FindActionIdInActionList(uint16_t actionId)
+{
+    if (GetInstance()->mDelegate == nullptr)
+    {
+        ChipLogError(Zcl, "Actions delegate is null!!!");
+        return false;
+    }
+    return GetInstance()->mDelegate->FindActionIdInActionList(actionId);
+}
+
+template <typename RequestT, typename FuncT>
+void Instance::HandleCommand(HandlerContext & handlerContext, FuncT func)
```

Comment:
Worth documenting why you can't just use CommandHandlerInterface::HandleCommand (because you need to do the FindActionIdInActionList check, right?)

Reply:
Right, done.